include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -327,83 +328,80 @@ __parallel_transform_reduce(_ExecutionPolicy&& __exec, _Up __u, _Cp __combine, _
+    _Size __n_groups = (__n - 1) / __size_per_work_group + 1; // number of work groups
+    _Size __n_items = (__n - 1) / __iters_per_work_item + 1; // number of bunch of elements that the algorithm processes
 
-    _PRINT_INFO_IN_DEBUG_MODE(__exec, __wgroup_size, __mcu);
+    _PRINT_INFO_IN_DEBUG_MODE(__exec, __work_group_size);
 
     // Create temporary global buffers to store temporary values
-    auto __temp_1 = sycl::buffer<_Tp>(sycl::range<1>(__n_groups));
-    auto __temp_2 = sycl::buffer<_Tp>(sycl::range<1>(__n_groups));
+    sycl::buffer<_Tp> __temp(sycl::range<1>(2 * __n_groups));
     // __is_first == true. Reduce over each work_group
     // __is_first == false. Reduce between work groups
     bool __is_first = true;
-    auto __buf_1_ptr = &__temp_1; // __buf_1_ptr is not accessed on the device when __is_first == true
-    auto __buf_2_ptr = &__temp_1;
+
+    // For memory utilization it's better to use one big buffer instead of two small because size of the buffer is close to a few MB
+    _Size __offset_1 = 0;
+    _Size __offset_2 = __n_groups;
+
     sycl::event __reduce_event;
     do
     {
         __reduce_event = __exec.queue().submit([&](sycl::handler& __cgh) {
             __cgh.depends_on(__reduce_event);
 
             oneapi::dpl::__ranges::__require_access(__cgh, __rngs...); //get an access to data under SYCL buffer
-            auto __temp_1_acc = __buf_1_ptr->template get_access<access_mode::read_write>(__cgh);
-            auto __temp_2_acc = __buf_2_ptr->template get_access<access_mode::write>(__cgh);
+            auto __temp_acc = __temp.template get_access<access_mode::read_write>(__cgh);
             sycl::accessor<_Tp, 1, access_mode::read_write, sycl::access::target::local> __temp_local(
-                sycl::range<1>(__wgroup_size), __cgh);
+                sycl::range<1>(__work_group_size), __cgh);
             __cgh.parallel_for<__kernel_name_t>(
 #if _ONEDPL_COMPILE_KERNEL
                 __kernel,
 #endif
-                sycl::nd_range<1>(sycl::range<1>(__n_groups * __wgroup_size), sycl::range<1>(__wgroup_size)),
+                sycl::nd_range<1>(sycl::range<1>(__n_groups * __work_group_size), sycl::range<1>(__work_group_size)),
                 [=](sycl::nd_item<1> __item_id) {
-                    auto __global_idx = __item_id.get_global_id(0);
-                    auto __local_idx = __item_id.get_local_id(0);
+                    ::std::size_t __global_idx = __item_id.get_global_id(0);
+                    ::std::size_t __local_idx = __item_id.get_local_id(0);
                     // 1. Initialization (transform part). Fill local memory
                     if (__is_first)
                     {
-                        __u(__item_id, __global_idx, __n, __temp_local, __rngs...);
+                        __u(__item_id, __n, __iters_per_work_item, __temp_local, __rngs...);
                     }
                     else
```

Comment:
Seems you handle `__iters_per_work_item` element sequentially only within `if (__is_first)` for the first run. Have you tried applying the same optimization here (for the next runs)?

I mean the following:
1. Reduce `__work_group_size` size, say 4 times, that means that `__iters_per_work_items = 4`
2. Preserve the same local memory size, so that it is `__iters_per_work_items * __work_group__size` elements
3. Fill local memory
4. Reduce `__iters_per_work_items` elements in each work item
5. Perform 2nd stage.

Work groups on the second and the next submits may be too lightweight, so that dispatching may cost more than execution.